python-package/securexgboost/remote_server.py

```diff
@@ -68,7 +72,312 @@ def is_ready(self):
         return True
 
     def invoke(self):
-        self._ret = self._func(self._params, self._usernames, self._signatures, self._sig_lengths)
+        node_ips = globals().get("nodes")
+        if not node_ips:
+            self._ret = self._func(self._params, self._usernames, self._signatures, self._sig_lengths)
+        else: # We're the RPC orchestrator
+            seq_num = self._seq_num
+            signers = self._usernames
+            signatures = self._signatures
+            sig_lengths = self._sig_lengths
+            channels = []
+            for channel_addr in node_ips:
+                channels.append(grpc.insecure_channel(channel_addr))
+        
+            # Store futures in a list
+            # Futures hold the result of asynchronous calls to each gRPC server
+            futures = []
+        
+            for channel in channels:
+                stub = remote_pb2_grpc.RemoteStub(channel)
+        
+                # Asynchronous calls to start job on each node
+                if self._func == rabit_remote_api.RabitInit:
+                    response_future = stub.rpc_RabitInit.future(remote_pb2.RabitParams(
+                        seq_num=seq_num,
```

Comment:
All the function messages here essentially take 5 arguments -- the actual params, seq_num, signer, signatures, and sig_lengths. I wonder if we can consolidate all the code? I think all that might require is to rename the params argument to be the same.

Reply:
I thought about this but I'm not sure we can do what you're saying. The `remote_pb2` message type is dependent on the invoked function, e.g. `XGDMatrixCreateFromEncryptedFile` requires `remote_pb2.DMatrixAttrsRequest` while `XGBoosterSetParam` requires `BoosterParamRequest`.  

Am I missing something?

Reply:
Ah no, you're right, I missed that.